lib/msprime.c

```diff
@@ -6375,19 +6367,38 @@ msp_set_simulation_model_beta(msp_t *self, double reference_size, double alpha,
         double truncation_point)
 {
     int ret = 0;
+    double m;
+    double scalar;
+
+    /* Numerical instability occurs for values above 2 - 0.0005. */
+    if (alpha <= 1.0 || alpha >= 2.0) {
+        ret = MSP_ERR_BAD_ALPHA;
+        goto out;
+    }
+
+    if (truncation_point <= 0.0 || truncation_point > 1.0) {
+        ret = MSP_ERR_BAD_TRUNCATION_POINT;
+        goto out;
+    }
 
-    /* TODO bounds check alpha and truncation_point: what are legal values? */
     ret = msp_set_simulation_model(self, MSP_MODEL_BETA, reference_size);
     if (ret != 0) {
         goto out;
     }
+
+    m = 2.0 + exp(alpha * 0.6931472 + (1 - alpha) * 1.098612 - log(alpha - 1));
+    scalar = exp(log(alpha) - alpha * log(m)
+             + gsl_sf_beta_inc(2 - alpha, alpha, truncation_point)
+             - (alpha - 1) * reference_size);
```

Comment:
I think this is a mistake inherited from earlier versions, but this doesn't look quite right. The RHS should be `exp(log(alpha) - alpha * log(m) - (alpha - 1) * log(reference_size)) * gsl_sf_beta_inc(2 - alpha, alpha)`.

Reply:
I think It would also make sense to change all the hard-coded decimal expressions to their respective exact formulas, e.g. `0.6931472` is really `log(2)`, `1.098612` is really `log(3)`, and so on. I think `log(4)` appears somewhere similarly too. I'm not sure why they've been written this way as it is, but if we want this to be stable for the greatest possible range of alphas, then they should be as accurate as (reasonably) possible. I think the `log()` versions are also more readable.

Reply:
Totally agree on putting in log(2) etc @JereKoskela, putting in constants is a premature optimisation (and an unreadable one).

Reply:
Agreed regarding the log()s. The log(4) is hiding inside `beta_integrand()`. If optimisation is really important (I doubt it in this case), we could add our own #defines for log(3) and log(4) (math.h has M_LOG2E for log(2)).

Reply:
FYI, log is usually implemented as a lookup table, and I'm guessing the compiler can inline the function call. So, I'd be surprised if log(constant) required any explicit optimisation.

Reply:
Totally agree @grahamgower, any compiler worth its salt is going to pre-compute these.